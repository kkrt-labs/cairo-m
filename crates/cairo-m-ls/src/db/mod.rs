//! Language Server database definitions and implementations.
//!
//! This module defines the concrete database used by the language server,
//! implementing all necessary traits from the compiler phases.

mod swapper;

use std::collections::HashMap;
use std::path::PathBuf;

use cairo_m_compiler_parser::{Db as ParserDb, SourceFile, Upcast};
use cairo_m_compiler_semantic::SemanticDb;
pub use swapper::AnalysisDatabaseSwapper;

/// The unified project crate representation used by the language server.
///
/// This serves as the single source of truth for project structure,
/// replacing the various ad-hoc crate creations throughout the codebase.
#[salsa::input(debug)]
pub struct ProjectCrate {
    /// The root directory of the project
    #[return_ref]
    pub root_dir: PathBuf,

    /// The main module name (e.g., "main" for main.cm)
    #[return_ref]
    pub main_module: String,

    /// All source files in the project, keyed by their absolute path
    #[return_ref]
    pub files: HashMap<PathBuf, SourceFile>,
}

/// The language server's analysis database.
///
/// This database extends the compiler database with additional
/// functionality needed for language server operations.
#[salsa::db]
#[derive(Clone, Default)]
pub struct AnalysisDatabase {
    storage: salsa::Storage<Self>,
}

impl AnalysisDatabase {
    /// Create a new analysis database.
    pub fn new() -> Self {
        Self::default()
    }
}

// Implement all required database traits
impl salsa::Database for AnalysisDatabase {}

#[salsa::db]
impl ParserDb for AnalysisDatabase {}

#[salsa::db]
impl SemanticDb for AnalysisDatabase {}

// Implement upcast traits for each database level
impl Upcast<dyn ParserDb> for AnalysisDatabase {
    fn upcast(&self) -> &(dyn ParserDb + 'static) {
        self
    }
    fn upcast_mut(&mut self) -> &mut (dyn ParserDb + 'static) {
        self
    }
}

impl Upcast<dyn SemanticDb> for AnalysisDatabase {
    fn upcast(&self) -> &(dyn SemanticDb + 'static) {
        self
    }
    fn upcast_mut(&mut self) -> &mut (dyn SemanticDb + 'static) {
        self
    }
}

/// Trait for converting ProjectCrate to the various crate representations
/// used by different compiler phases.
pub trait ProjectCrateExt {
    /// Convert to a semantic::Crate for semantic analysis.
    fn to_semantic_crate(&self, db: &dyn SemanticDb) -> cairo_m_compiler_semantic::Crate;
}

// Module path resolution is now centralized in cairo-m-project crate.
// This implementation delegates to the compiler's canonical module naming logic.
impl ProjectCrateExt for ProjectCrate {
    fn to_semantic_crate(&self, db: &dyn SemanticDb) -> cairo_m_compiler_semantic::Crate {
        let files = self.files(db);
        let main_module = self.main_module(db);
        let root_dir = self.root_dir(db);

        let mut modules = HashMap::new();

        // Create a temporary Project instance to use its module_name_from_path method
        let project = cairo_m_project::Project {
            manifest_path: root_dir.join("cairom.toml"),
            root_directory: root_dir.clone(),
            name: root_dir
                .file_name()
                .and_then(|n| n.to_str())
                .unwrap_or("unknown")
                .to_string(),
            source_layout: cairo_m_project::SourceLayout::default(),
            entry_point: None,
        };

        // Convert PathBuf keys to module names with proper nesting
        for (path, source_file) in files {
            // Use the Project's module_name_from_path method
            match project.module_name_from_path(&path) {
                Ok(module_name) => {
                    modules.insert(module_name, source_file);
                }
                Err(_) => {
                    // Fallback for files outside project root (shouldn't happen)
                    if let Some(module_name) = path.file_stem().and_then(|s| s.to_str()) {
                        modules.insert(module_name.to_string(), source_file);
                    }
                }
            }
        }

        cairo_m_compiler_semantic::Crate::new(
            db,
            modules,
            main_module,
            root_dir.clone(),
            root_dir
                .file_name()
                .and_then(|n| n.to_str())
                .unwrap_or("unknown")
                .to_string(),
        )
    }
}
