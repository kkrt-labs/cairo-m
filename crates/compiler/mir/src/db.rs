//! Database traits and implementations for MIR generation with Salsa integration.

use cairo_m_compiler_parser::{SourceProgram, Upcast};
use cairo_m_compiler_semantic::SemanticDb;

use crate::{MirError, MirModule};

/// Database trait for MIR (Mid-level Intermediate Representation) queries.
///
/// This trait extends SemanticDb to provide MIR-specific queries that are cached
/// and incrementally recomputed by Salsa. All MIR generation should go through
/// these queries to benefit from incremental compilation.
#[salsa::db]
pub trait MirDb: SemanticDb + Upcast<dyn SemanticDb> {}

/// Generate MIR for a source file.
///
/// This is the main entry point for MIR generation. It uses the semantic index
/// to build the MIR module, with full incremental caching support.
#[salsa::tracked]
pub fn generate_mir(db: &dyn MirDb, file: SourceProgram) -> Option<MirModule> {
    // Get the parsed module
    let parsed = cairo_m_compiler_parser::parse_program(db.upcast(), file);

    if !parsed.diagnostics.is_empty() {
        // Don't generate MIR if there are parse errors
        return None;
    }

    // Delegate to the existing generate_mir function from ir_generation
    crate::ir_generation::generate_mir(db, file).map(|arc| (*arc).clone())
}

/// Track MIR-specific errors separately for better diagnostics.
///
/// This allows us to report MIR generation errors without blocking
/// other compilation phases.
#[salsa::tracked]
pub fn mir_errors(_db: &dyn MirDb, _file: SourceProgram) -> Vec<MirError> {
    // TODO
    // For now, we'll return an empty vector
    // In the future, this should collect errors from MIR generation
    vec![]
}

#[cfg(test)]
pub(crate) mod tests {
    use super::*;

    /// Test database that implements all required traits for MIR generation
    #[salsa::db]
    #[derive(Clone, Default)]
    pub struct TestDatabase {
        storage: salsa::Storage<Self>,
    }

    #[salsa::db]
    impl salsa::Database for TestDatabase {}

    #[salsa::db]
    impl cairo_m_compiler_parser::Db for TestDatabase {}

    #[salsa::db]
    impl SemanticDb for TestDatabase {}

    #[salsa::db]
    impl MirDb for TestDatabase {}

    impl Upcast<dyn cairo_m_compiler_parser::Db> for TestDatabase {
        fn upcast(&self) -> &(dyn cairo_m_compiler_parser::Db + 'static) {
            self
        }
        fn upcast_mut(&mut self) -> &mut (dyn cairo_m_compiler_parser::Db + 'static) {
            self
        }
    }

    impl Upcast<dyn SemanticDb> for TestDatabase {
        fn upcast(&self) -> &(dyn SemanticDb + 'static) {
            self
        }
        fn upcast_mut(&mut self) -> &mut (dyn SemanticDb + 'static) {
            self
        }
    }

    pub fn test_db() -> TestDatabase {
        TestDatabase::default()
    }

    #[test]
    fn test_mir_db_trait() {
        let db = TestDatabase::default();
        let source = SourceProgram::new(&db, "fn main() {}".to_string(), "test.cm".to_string());

        // This should trigger MIR generation through Salsa
        let _mir = generate_mir(&db, source);
    }
}
