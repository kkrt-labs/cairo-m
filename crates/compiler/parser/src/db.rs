#[salsa::db]
#[derive(Clone, Default)]
pub struct ParserDatabaseImpl {
    storage: salsa::Storage<Self>,
}

#[salsa::db]
impl salsa::Database for ParserDatabaseImpl {}
