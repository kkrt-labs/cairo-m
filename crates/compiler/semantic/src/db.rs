//! # Semantic Analysis Database
//!
//! This module defines the database traits and implementations for semantic analysis
//! using the Salsa incremental computation framework. It extends the parser database
//! to provide semantic-specific functionality.
//!
//! The database system enables incremental recompilation by caching query results
//! and invalidating them only when their dependencies change.

use cairo_m_compiler_diagnostics::DiagnosticCollection;
use cairo_m_compiler_parser as parser;
// We may need the parser implementation later for cross-database operations
#[allow(unused_imports)]
use parser::ParserDatabaseImpl;
use parser::{Db as ParserDb, Upcast};

/// Database trait for semantic analysis, extending the parser database
///
/// This trait defines the interface for semantic-specific database operations.
/// Type resolution and inference queries are defined as standalone tracked functions.
#[salsa::db]
pub trait SemanticDb: ParserDb + Upcast<dyn ParserDb> {
    // Type queries are defined as standalone tracked functions in type_resolution.rs
    // This trait includes any database-specific configuration or settings
}

/// Concrete database implementation for semantic analysis
///
/// This provides the actual storage and implementation for all database queries.
/// It combines both parser and semantic analysis capabilities in a single database.
///
/// # Thread Safety
///
/// This implementation is `Clone` and can be safely shared between threads.
/// Salsa handles the synchronization internally.
#[salsa::db]
#[derive(Clone, Default)]
pub struct SemanticDatabaseImpl {
    storage: salsa::Storage<Self>,
}

impl salsa::Database for SemanticDatabaseImpl {}
#[salsa::db]
impl ParserDb for SemanticDatabaseImpl {}
#[salsa::db]
impl SemanticDb for SemanticDatabaseImpl {}

impl Upcast<dyn ParserDb> for SemanticDatabaseImpl {
    fn upcast(&self) -> &(dyn ParserDb + 'static) {
        self
    }
    fn upcast_mut(&mut self) -> &mut (dyn ParserDb + 'static) {
        self
    }
}

/// Validate semantics of a source file and return diagnostics.
///
/// This is a tracked query that performs comprehensive semantic validation
/// including scope checking, type checking, and control flow analysis.
#[salsa::tracked]
pub fn validate_semantics(
    db: &dyn SemanticDb,
    file: parser::SourceProgram,
) -> DiagnosticCollection {
    // Parse the program first
    let parsed = parser::parse_program(db.upcast(), file);

    // Get the semantic index (this is already a tracked query)
    let index = crate::semantic_index::semantic_index_from_module(db, &parsed.module, file);

    // Create validator registry with all available validators
    let registry = crate::validation::validator::create_default_registry();
    registry.validate_all(db, file, &index)
}

#[cfg(test)]
pub(crate) mod tests {
    use super::*;

    #[salsa::db]
    #[derive(Clone)]
    pub struct TestDb {
        storage: salsa::Storage<Self>,
    }

    impl salsa::Database for TestDb {}
    #[salsa::db]
    impl ParserDb for TestDb {}
    #[salsa::db]
    impl SemanticDb for TestDb {}

    impl Upcast<dyn ParserDb> for TestDb {
        fn upcast(&self) -> &(dyn ParserDb + 'static) {
            self
        }
        fn upcast_mut(&mut self) -> &mut (dyn ParserDb + 'static) {
            self
        }
    }

    pub fn test_db() -> TestDb {
        TestDb {
            storage: salsa::Storage::default(),
        }
    }
}
