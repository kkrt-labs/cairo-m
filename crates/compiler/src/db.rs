//! Unified database implementation for the Cairo-M compiler.
//!
//! This module provides a single database that implements all compilation phases
//! (parsing, semantic analysis, MIR generation, and code generation) with full
//! Salsa incremental compilation support.

use cairo_m_compiler_codegen::CodegenDb;
use cairo_m_compiler_mir::MirDb;
use cairo_m_compiler_parser::{Db as ParserDb, Upcast};
use cairo_m_compiler_semantic::SemanticDb;

/// The main compiler database that supports all compilation phases.
///
/// This database implements all the trait requirements for:
/// - Parsing (ParserDb)
/// - Semantic analysis (SemanticDb)
/// - MIR generation (MirDb)
/// - Code generation (CodegenDb)
///
/// It provides a unified interface for the entire compilation pipeline
/// with automatic incremental compilation support through Salsa.
#[salsa::db]
#[derive(Clone, Default)]
pub struct CompilerDatabase {
    storage: salsa::Storage<Self>,
}

// Implement all required database traits
impl salsa::Database for CompilerDatabase {}

#[salsa::db]
impl ParserDb for CompilerDatabase {}

#[salsa::db]
impl SemanticDb for CompilerDatabase {}

#[salsa::db]
impl MirDb for CompilerDatabase {}

#[salsa::db]
impl CodegenDb for CompilerDatabase {}

// Implement upcast traits for each database level
impl Upcast<dyn ParserDb> for CompilerDatabase {
    fn upcast(&self) -> &(dyn ParserDb + 'static) {
        self
    }
    fn upcast_mut(&mut self) -> &mut (dyn ParserDb + 'static) {
        self
    }
}

impl Upcast<dyn SemanticDb> for CompilerDatabase {
    fn upcast(&self) -> &(dyn SemanticDb + 'static) {
        self
    }
    fn upcast_mut(&mut self) -> &mut (dyn SemanticDb + 'static) {
        self
    }
}

impl Upcast<dyn MirDb> for CompilerDatabase {
    fn upcast(&self) -> &(dyn MirDb + 'static) {
        self
    }
    fn upcast_mut(&mut self) -> &mut (dyn MirDb + 'static) {
        self
    }
}

impl Upcast<dyn CodegenDb> for CompilerDatabase {
    fn upcast(&self) -> &(dyn CodegenDb + 'static) {
        self
    }
    fn upcast_mut(&mut self) -> &mut (dyn CodegenDb + 'static) {
        self
    }
}

impl CompilerDatabase {
    /// Create a new compiler database.
    pub fn new() -> Self {
        Self::default()
    }
}
