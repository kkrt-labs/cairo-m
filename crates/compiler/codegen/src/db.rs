//! Database traits and implementations for code generation with Salsa integration.

use std::sync::Arc;

use cairo_m_common::Program;
use cairo_m_compiler_mir::{MirDb, MirModule};
use cairo_m_compiler_parser::Upcast;
use cairo_m_compiler_semantic::db::Crate;

use crate::CodegenError;

/// Database trait for code generation queries.
///
/// This trait extends MirDb to provide code generation queries that are cached
/// and incrementally recomputed by Salsa. All code generation should go through
/// these queries to benefit from incremental compilation.
#[salsa::db]
pub trait CodegenDb: MirDb + Upcast<dyn MirDb> {}

/// Compile a crate to a compiled program.
///
/// This is the main entry point for code generation. It takes a crate
/// and produces the compiled program with full incremental caching support.
#[salsa::tracked]
pub fn compile_project(db: &dyn CodegenDb, crate_id: Crate) -> Result<Arc<Program>, CodegenError> {
    // Get the MIR module
    let mir_module = cairo_m_compiler_mir::db::generate_mir(db.upcast(), crate_id)
        .ok_or_else(|| CodegenError::InvalidMir("No MIR module generated".to_string()))?;

    // Use the existing compile_module logic
    let compiled = crate::compile_module(&mir_module)?;

    Ok(Arc::new(compiled))
}

/// Get the MIR module for a crate (convenience re-export).
///
/// This allows code generation to access MIR without directly depending
/// on the MIR crate's internals.
#[salsa::tracked]
pub fn codegen_mir_module(db: &dyn CodegenDb, crate_id: Crate) -> Option<Arc<MirModule>> {
    cairo_m_compiler_mir::db::generate_mir(db.upcast(), crate_id).map(Arc::new)
}

/// Track code generation errors separately for better diagnostics.
///
/// This allows us to report codegen errors without blocking other phases.
#[salsa::tracked]
pub fn codegen_errors(db: &dyn CodegenDb, crate_id: Crate) -> Vec<CodegenError> {
    // Collect errors from code generation
    match compile_project(db, crate_id) {
        Ok(_) => vec![],
        Err(e) => vec![e],
    }
}

#[cfg(test)]
pub(crate) mod tests {
    use std::collections::HashMap;
    use std::path::PathBuf;

    use cairo_m_compiler_mir::MirDb;
    use cairo_m_compiler_semantic::{File, SemanticDb};

    use super::*;

    /// Test database that implements all required traits for code generation
    #[salsa::db]
    #[derive(Clone, Default)]
    pub struct TestDatabase {
        storage: salsa::Storage<Self>,
    }

    #[salsa::db]
    impl salsa::Database for TestDatabase {}

    #[salsa::db]
    impl cairo_m_compiler_parser::Db for TestDatabase {}

    #[salsa::db]
    impl SemanticDb for TestDatabase {}

    #[salsa::db]
    impl MirDb for TestDatabase {}

    #[salsa::db]
    impl CodegenDb for TestDatabase {}

    impl Upcast<dyn cairo_m_compiler_parser::Db> for TestDatabase {
        fn upcast(&self) -> &(dyn cairo_m_compiler_parser::Db + 'static) {
            self
        }
        fn upcast_mut(&mut self) -> &mut (dyn cairo_m_compiler_parser::Db + 'static) {
            self
        }
    }

    impl Upcast<dyn SemanticDb> for TestDatabase {
        fn upcast(&self) -> &(dyn SemanticDb + 'static) {
            self
        }
        fn upcast_mut(&mut self) -> &mut (dyn SemanticDb + 'static) {
            self
        }
    }

    impl Upcast<dyn MirDb> for TestDatabase {
        fn upcast(&self) -> &(dyn MirDb + 'static) {
            self
        }
        fn upcast_mut(&mut self) -> &mut (dyn MirDb + 'static) {
            self
        }
    }

    #[test]
    fn test_codegen_db_trait() {
        let db = TestDatabase::default();
        let file = File::new(&db, "fn main() {}".to_string(), "test.cm".to_string());
        let mut modules = HashMap::new();
        modules.insert("main".to_string(), file);
        let crate_id = Crate::new(
            &db,
            modules,
            "main".to_string(),
            PathBuf::from("."),
            "test".to_string(),
        );

        // This should trigger code generation through Salsa
        let _result = compile_project(&db, crate_id);
    }
}
