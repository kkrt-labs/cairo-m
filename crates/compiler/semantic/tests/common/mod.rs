//! Common test utilities for semantic analysis tests
//!
//! This module contains all shared test infrastructure including:
//! - Test database setup
//! - Crate creation utilities
//! - Helper functions for semantic index access
//! - Test code generation helpers
//! - Diagnostic formatting and assertion macros

use std::collections::HashMap;
use std::path::PathBuf;

use cairo_m_compiler_diagnostics::{
    DiagnosticCode, DiagnosticCollection, build_diagnostic_message,
};
use cairo_m_compiler_parser::{Db as ParserDb, SourceFile, Upcast};
use cairo_m_compiler_semantic::db::{Crate, project_validate_semantics};
use cairo_m_compiler_semantic::{File, SemanticDb, SemanticIndex, project_semantic_index};
use insta::assert_snapshot;

// ===== Test Database Setup =====

#[salsa::db]
#[derive(Clone, Default)]
pub struct TestDb {
    storage: salsa::Storage<Self>,
}

#[salsa::db]
impl salsa::Database for TestDb {}
#[salsa::db]
impl ParserDb for TestDb {}
#[salsa::db]
impl SemanticDb for TestDb {}

impl Upcast<dyn ParserDb> for TestDb {
    fn upcast(&self) -> &(dyn ParserDb + 'static) {
        self
    }
    fn upcast_mut(&mut self) -> &mut (dyn ParserDb + 'static) {
        self
    }
}

pub fn test_db() -> TestDb {
    TestDb::default()
}

// ===== Test Project Utilities =====

/// Represents a test project that can be either a single file or multiple files
#[derive(Debug, Clone)]
pub enum TestProject {
    /// A test case with a single source file.
    Single(String),
    /// A test case representing a multi-file crate.
    Multi {
        /// The name of the entry-point module (e.g., "main.cm").
        main_module: String,
        /// A vector of (file_name, file_content) tuples.
        modules: Vec<(String, String)>,
    },
}

/// A helper to conveniently create a multi-file test project.
pub fn multi_file<'a>(main_module: &'a str, modules: &'a [(&'a str, &'a str)]) -> TestProject {
    TestProject::Multi {
        main_module: main_module.to_string(),
        modules: modules
            .iter()
            .map(|(name, content)| {
                (
                    name.split("/").collect::<Vec<&str>>().join("::"),
                    content.to_string(),
                )
            })
            .collect(),
    }
}

/// Allows existing `&str` test cases to be converted automatically.
impl<'a> From<&'a str> for TestProject {
    fn from(code: &'a str) -> Self {
        Self::Single(code.to_string())
    }
}

/// Allows existing String test cases to be converted automatically.
impl From<String> for TestProject {
    fn from(code: String) -> Self {
        Self::Single(code)
    }
}

// ===== Crate Creation Utilities =====

pub fn single_file_crate(db: &dyn SemanticDb, file: File) -> Crate {
    let mut modules = HashMap::new();
    modules.insert("main".to_string(), file);
    Crate::new(
        db,
        modules,
        "main".to_string(),
        PathBuf::from("."),
        "crate_test".to_string(),
    )
}

pub fn crate_from_program(db: &dyn SemanticDb, program: &str) -> Crate {
    let file = File::new(db, program.to_string(), "test.cm".to_string());
    single_file_crate(db, file)
}

pub fn get_main_semantic_index(db: &dyn SemanticDb, crate_id: Crate) -> SemanticIndex {
    let semantic_index = project_semantic_index(db, crate_id).unwrap();
    semantic_index.modules().values().next().unwrap().clone()
}

// ===== Validation and Diagnostic Utilities =====

/// Run semantic validation on source code
pub fn run_validation(source: &str, file_name: &str) -> DiagnosticCollection {
    let db = test_db();
    let source_program = SourceFile::new(&db, source.to_string(), file_name.to_string());
    let crate_id = single_file_crate(&db, source_program);
    project_validate_semantics(&db, crate_id)
}

// Format diagnostics for snapshot testing using ariadne for beautiful error reports
pub fn format_diagnostics_for_snapshot(
    diagnostics: &DiagnosticCollection,
    source: &str,
    fixture_name: &str,
) -> String {
    let mut result = String::new();

    // Add header
    result.push_str(&format!("Fixture: {fixture_name}\n"));
    result.push_str(&"=".repeat(60));
    result.push('\n');
    result.push_str(&format!("Source code:\n{source}\n"));
    result.push_str(&"=".repeat(60));
    result.push('\n');

    if diagnostics.is_empty() {
        result.push_str("No diagnostics found.\n");
        return result;
    }

    result.push_str(&format!("Found {} diagnostic(s):\n\n", diagnostics.len()));

    // Create ariadne reports for each diagnostic
    for (i, diagnostic) in diagnostics.all().iter().enumerate() {
        let message = build_diagnostic_message(source, diagnostic, false);
        result.push_str(&format!("--- Diagnostic {} ---\n", i + 1));
        result.push_str(&message);
        if i < diagnostics.len() - 1 {
            result.push('\n');
        }
    }

    result
}

// ===== Assertion Implementation Functions =====

/// Assert that inline code validates successfully without any diagnostics
/// Similar to parser's assert_parses_ok! but for semantic validation
#[track_caller]
pub fn assert_semantic_ok_impl(code: &str, test_name: &str) {
    assert_semantic_ok_impl_with_options(code, test_name, false)
}

/// Assert that inline code validates successfully without any diagnostics
/// with option to mute unused variable warnings
#[track_caller]
pub fn assert_semantic_ok_impl_with_options(
    code: &str,
    test_name: &str,
    show_unused_warnings: bool,
) {
    let diagnostics = run_validation(code, test_name);
    let filtered_diagnostics = if show_unused_warnings {
        diagnostics
    } else {
        filter_unused_variable_warnings(&diagnostics)
    };

    if !filtered_diagnostics.is_empty() {
        let report = format_diagnostics_for_snapshot(&filtered_diagnostics, code, test_name);
        panic!("Expected successful semantic validation, but got diagnostics:\n{report}");
    }
}

/// Assert that inline code fails semantic validation and produces diagnostics
/// Similar to parser's assert_parses_err! but for semantic validation
#[track_caller]
pub fn assert_semantic_err_impl(code: &str, test_name: &str) {
    assert_semantic_err_impl_with_options(code, test_name, false)
}

/// Assert that inline code fails semantic validation and produces diagnostics
/// with option to mute unused variable warnings
#[track_caller]
pub fn assert_semantic_err_impl_with_options(
    code: &str,
    test_name: &str,
    show_unused_warnings: bool,
) {
    let diagnostics = run_validation(code, test_name);
    let filtered_diagnostics = if show_unused_warnings {
        diagnostics
    } else {
        filter_unused_variable_warnings(&diagnostics)
    };

    if filtered_diagnostics.is_empty() {
        panic!("Expected semantic validation to fail, but it succeeded without diagnostics.");
    }

    let snapshot_content = format_diagnostics_for_snapshot(&filtered_diagnostics, code, test_name);

    // Extract local test name for better snapshot organization
    let base_path = "semantic_tests::";
    let local_test_name = test_name.split(base_path).nth(1).unwrap_or(test_name);

    insta::with_settings!({
        description => format!("Inline semantic validation error test: {}", local_test_name),
        omit_expression => true,
        sort_maps => true,
        prepend_module_to_snapshot => false,
    }, {
        assert_snapshot!(format!("diagnostics__{}", local_test_name), snapshot_content);
    });
}

/// Helper to wrap statement code inside a function, since most statements are not top-level.
pub fn in_function(code: &str) -> String {
    format!("fn test() {{ {code} return; }}")
}

/// Filter out unused variable warnings from diagnostics
pub fn filter_unused_variable_warnings(diagnostics: &DiagnosticCollection) -> DiagnosticCollection {
    let filtered: Vec<_> = diagnostics
        .all()
        .iter()
        .filter(|d| d.code != DiagnosticCode::UnusedVariable)
        .cloned()
        .collect();
    DiagnosticCollection::from(filtered)
}

// ===== New Parameterized Testing Infrastructure =====

/// Result of parameterized semantic tests
#[derive(Debug)]
pub struct ParameterizedSemanticResults {
    pub results: Vec<ParameterizedSemanticResult>,
}

#[derive(Debug)]
pub enum ParameterizedSemanticResult {
    Error {
        input: String,
        diagnostics: DiagnosticCollection,
    },
}

impl std::fmt::Display for ParameterizedSemanticResults {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        for (i, result) in self.results.iter().enumerate() {
            if i > 0 {
                writeln!(f, "\n{}\n", "=".repeat(60))?;
            }

            match result {
                ParameterizedSemanticResult::Error { input, diagnostics } => {
                    writeln!(f, "--- Input {} (ERROR) ---", i + 1)?;
                    writeln!(f, "{}", input)?;
                    writeln!(f, "--- Diagnostics ---")?;
                    for diagnostic in diagnostics.all() {
                        let message = build_diagnostic_message(input, diagnostic, false);
                        write!(f, "{}", message)?;
                    }
                }
            }
        }
        Ok(())
    }
}

/// Assert that multiple code snippets validate with expected results
#[track_caller]
pub fn assert_semantic_parameterized_impl(
    inputs: &[(TestProject, bool)], // (project, should_succeed)
    test_name: &str,
    show_unused_warnings: bool,
) {
    let mut results = Vec::new();

    for (index, (project, should_succeed)) in inputs.iter().enumerate() {
        let db = test_db();

        let (diagnostics, input_str) = match project {
            TestProject::Single(code) => {
                // Create a single-file crate and validate.
                let source_file = SourceFile::new(&db, code.clone(), test_name.to_string());
                let crate_id = single_file_crate(&db, source_file);
                let diagnostics = project_validate_semantics(&db, crate_id);
                (diagnostics, code.clone())
            }
            TestProject::Multi {
                main_module,
                modules,
            } => {
                let mut module_map = HashMap::new();
                let mut all_content = String::new();

                for (name, content) in modules {
                    let module_name = name.strip_suffix(".cm").unwrap_or(name);
                    let source_file = SourceFile::new(&db, content.clone(), name.clone());
                    module_map.insert(module_name.to_string(), source_file);

                    // Collect content for error display
                    if !all_content.is_empty() {
                        all_content.push_str("\n// --- ");
                        all_content.push_str(name);
                        all_content.push_str(" ---\n");
                    }
                    all_content.push_str(content);
                }

                let main_module_name = main_module.strip_suffix(".cm").unwrap_or(main_module);
                let crate_id = Crate::new(
                    &db,
                    module_map,
                    main_module_name.to_string(),
                    PathBuf::from("."),
                    "test_crate".to_string(),
                );
                let diagnostics = project_validate_semantics(&db, crate_id);
                (diagnostics, all_content)
            }
        };

        let filtered_diagnostics = if show_unused_warnings {
            diagnostics
        } else {
            filter_unused_variable_warnings(&diagnostics)
        };

        if *should_succeed {
            if !filtered_diagnostics.is_empty() {
                let report =
                    format_diagnostics_for_snapshot(&filtered_diagnostics, &input_str, test_name);
                panic!(
                    "Expected successful semantic validation for input {} (index {}), but got diagnostics:\n{}",
                    match project {
                        TestProject::Single(_) => "single file".to_string(),
                        TestProject::Multi { main_module, .. } =>
                            format!("multi-file with main '{}'", main_module),
                    },
                    index,
                    report
                );
            }
        } else {
            if filtered_diagnostics.is_empty() {
                panic!(
                    "Expected semantic validation to fail for input {} (index {}), but it succeeded",
                    match project {
                        TestProject::Single(_) => "single file".to_string(),
                        TestProject::Multi { main_module, .. } =>
                            format!("multi-file with main '{}'", main_module),
                    },
                    index
                );
            }
            results.push(ParameterizedSemanticResult::Error {
                input: input_str,
                diagnostics: filtered_diagnostics,
            });
        }
    }

    let snapshot = ParameterizedSemanticResults { results };

    let base_path = "semantic_tests::";
    let local_test_name = test_name.split(base_path).nth(1).unwrap_or(test_name);

    insta::with_settings!({
        prepend_module_to_snapshot => false,
    }, {
        assert_snapshot!(format!("parameterized__{}", local_test_name), snapshot);
    });
}
