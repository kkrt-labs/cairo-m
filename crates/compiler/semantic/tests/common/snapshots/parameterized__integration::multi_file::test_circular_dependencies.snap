---
source: crates/compiler/semantic/tests/common/mod.rs
expression: snapshot
---
--- Input 1 (ERROR) ---
use module_b::func_b;
fn func_a() { func_b(); }
// --- module_b.cm ---
use module_a::func_a;
fn func_b() { func_a(); }
--- Diagnostics ---
[02] Error: Cyclic import: module_a -> module_b -> module_a
   ╭─[ :1:1 ]
   │
 1 │ use module_b::func_b;
   │ │ 
   │ ╰─ Cyclic import: module_a -> module_b -> module_a
───╯

============================================================

--- Input 2 (ERROR) ---
use module_b::func_b;
fn func_a() { func_b(); }
// --- module_b.cm ---
use module_c::func_c;
fn func_b() { func_c(); }
// --- module_c.cm ---
use module_a::func_a;
fn func_c() { func_a(); }
--- Diagnostics ---
[02] Error: Cyclic import: module_a -> module_b -> module_c -> module_a
   ╭─[ :1:1 ]
   │
 1 │ use module_b::func_b;
   │ │ 
   │ ╰─ Cyclic import: module_a -> module_b -> module_c -> module_a
───╯

============================================================

--- Input 3 (ERROR) ---
use main::foo;
fn foo() {}
--- Diagnostics ---
[02] Error: Cyclic import: main -> main
   ╭─[ :1:1 ]
   │
 1 │ use main::foo;
   │ │ 
   │ ╰─ Cyclic import: main -> main
───╯
